### PR TITLE
Introducing Workspaces into pipelines

### PR DESCRIPTION
**Pull request by bhanurp**
- [x] All [tests](https://github.com/jfrog/jfrog-client-go#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] All [static analysis checks](https://github.com/jfrog/jfrog-client-go/actions/workflows/analysis.yml) passed.
- [x] This pull request is on the dev branch.
- [x] I used gofmt for formatting the code before submitting the pull request.
-----

### CODE REVIEW

File: pipelines/services/workspace.go [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+}
+
+// GetStepLogsUsingStepID retrieve steps logs using step id
+func (ws *WorkspaceService) GetStepLogsUsingStepID(stepID string) (map[string][]Console, error) {
+	httpDetails := ws.getHttpDetails()
+	stepConsolesAPI := strings.Replace(stepConsoles, ":stepID", stepID, 1)
+	// Query params
+	queryParams := make(map[string]string, 0)
+	// URL construction
+	uri, err := constructPipelinesURL(queryParams, ws.ServiceDetails.GetUrl(), stepConsolesAPI)
```
Comment (yahavi) [inline pipelines/services/workspace.go:302]:
```suggestion
	uri, err := constructPipelinesURL(queryParams, ws.GetUrl(), stepConsolesAPI)
```

File: pipelines/services/workspace.go [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+	if err != nil {
+		return nil, err
+	}
+	// Response Analysis
+	if err = errorutils.CheckResponseStatusWithBody(resp, body, http.StatusOK); err != nil {
+		return nil, err
+	}
+	consoles := make(map[string][]Console)
+	err = json.Unmarshal(body, &consoles)
+	return consoles, err
```
Comment (yahavi) [inline pipelines/services/workspace.go:342]:
```suggestion
	return consoles, errorutils.CheckError(err)
```

File: tests/pipelinesworkspaces_test.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+)
+
+func TestPipelinesWorkspaceService(t *testing.T) {
+	t.Run("test trigger pipelines workspace when resources are not valid", TestWorkspaceValidationWhenPipelinesResourcesAreNotValid)
+	t.Run("test trigger pipelines workspace when resources are valid", TestWorkspaceValidationWhenPipelinesResourcesAreValid)
+	t.Run("test workspace sync when it fails for the first time and succeeds later", TestWorkspaceValidationFailureAndSucceedsWhenValidIntegrationCreated)
+}
+
+func TestWorkspaceValidationWhenPipelinesResourcesAreNotValid(t *testing.T) {
+
```
Comment (yahavi) [inline tests/pipelinesworkspaces_test.go:19]:
```suggestion
```

File: tests/pipelinesworkspaces_test.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+		return
+	}
+	filePath := "testdata/pipelines.yml"
+
+	// get pipelines.yaml content
+	pipelineRes, err := getWorkspaceRunPayload([]string{filePath})
+	if !assert.NoError(t, err) {
+		return
+	}
+	fmt.Printf("Bytes received : %s \n", pipelineRes)
```
Comment (yahavi) [inline tests/pipelinesworkspaces_test.go:30]:
```suggestion
```

File: tests/testdata/pip.yaml
Comment (yahavi) [inline tests/testdata/pip.yaml:1]: Let's avoid using the word "pip" for JFrog pipelines in this and other JFrog CLI components. You can name this file "pipe".

File: tests/pipelinesworkspaces_test.go [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+	"encoding/json"
+	"fmt"
+	"github.com/jfrog/jfrog-client-go/pipelines/services"
+	"github.com/stretchr/testify/assert"
+	"os"
+	"testing"
+)
+
+func TestPipelinesWorkspaceService(t *testing.T) {
+	t.Run("test trigger pipelines workspace when resources are not valid", TestWorkspaceValidationWhenPipelinesResourcesAreNotValid)
```
Comment (yahavi) [inline tests/pipelinesworkspaces_test.go:13]:
```suggestion
	initPipelinesTest(t)
	t.Run("test trigger pipelines workspace when resources are not valid", TestWorkspaceValidationWhenPipelinesResourcesAreNotValid)
```

**Review by yahavi — CHANGES_REQUESTED** [review]
@bhanurp 
Good work!
I reviewed your code and added some comments. I think a second round of code review is needed after fixing the issues.

Important - 
1. Ensure the tests are not failing
2. Add all new methods in manager.go to the README

File: pipelines/services/validate.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+	m := make(map[string]string, 0)
+
+	// URL Construction
+	uri := vs.constructValidateAPIURL(m, validatePipResourcePath)
+
+	// Headers
+	headers := make(map[string]string)
+	headers["Content-Type"] = "application/json"
+	httpDetails.Headers = headers
+	log.Debug(string(data))
```
Comment (yahavi) [inline pipelines/services/validate.go:49]:
```suggestion
```

File: pipelines/services/validate.go [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+	// Query params
+	m := make(map[string]string, 0)
+
+	// URL Construction
+	uri := vs.constructValidateAPIURL(m, validatePipResourcePath)
+
+	// Headers
+	headers := make(map[string]string)
+	headers["Content-Type"] = "application/json"
```
Comment (yahavi) [inline pipelines/services/validate.go:47]:
```suggestion
	utils.SetContentType("application/json", &headers)
```

File: pipelines/services/run.go [hunk 8]
```diff
@@ -99,19 +100,29 @@ func (rs *RunService) TriggerPipelineRun(branch, pipeline string, isMultiBranch
...
+		resp, body, err := rs.client.SendPost(uri, buf.Bytes(), &httpDetails)
+		if err != nil {
+			return true, body, err
+		}
+
+		// Response Analysis
+		if err := errorutils.CheckResponseStatusWithBody(resp, body, http.StatusOK); err != nil {
+			return false, body, err
+		}
+		log.Info(fmt.Sprintf("Triggered successfully\n%s %s \n%14s %s", "PipelineName :", pipeline, "Branch :", branch))
```
Comment (yahavi) [inline pipelines/services/run.go:114]:
```suggestion
		log.Info(fmt.Sprintf("Triggered successfully\n%s %s \n%14s %s", "PipelineName:", pipeline, "Branch:", branch))
```

File: pipelines/services/types.go [hunk 9]
```diff
@@ -87,3 +87,71 @@ type PipelineResources struct {
...
+	Timestamp        int64       `json:"timestamp,omitempty"`
+	TimestampEndedAt interface{} `json:"timestampEndedAt,omitempty"`
+	Type             string      `json:"type,omitempty"`
+	Message          string      `json:"message,omitempty"`
+	CreatedAt        time.Time   `json:"createdAt,omitempty"`
+	UpdatedAt        time.Time   `json:"updatedAt,omitempty"`
+}
+
+// Validation Types
+
```
Comment (yahavi) [inline pipelines/services/types.go:148]:
```suggestion
```

File: pipelines/services/validate.go [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+	client *jfroghttpclient.JfrogHttpClient
+	auth.ServiceDetails
+}
+
+func NewValidateService(client *jfroghttpclient.JfrogHttpClient) *ValidateService {
+	return &ValidateService{client: client}
+}
+
+func (vs *ValidateService) getHttpDetails() httputils.HttpClientDetails {
+	httpDetails := vs.ServiceDetails.CreateHttpClientDetails()
```
Comment (yahavi) [inline pipelines/services/validate.go:31]:
```suggestion
	httpDetails := vs.CreateHttpClientDetails()
```

File: pipelines/services/validate.go [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+	return httpDetails
+}
+
+func (vs *ValidateService) ValidatePipeline(data []byte) error {
+	var err error
+	httpDetails := vs.getHttpDetails()
+
+	// Query params
+	m := make(map[string]string, 0)
```
Comment (yahavi) [inline pipelines/services/validate.go:40]: What is "m"?

File: pipelines/services/validate.go [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+
+	// Process response
+	if nil != resp && resp.StatusCode == http.StatusOK {
+		log.Info("Processing resources yaml response ")
+		log.Info(string(body))
+		err = processValidatePipResourceResponse(body)
+		if err != nil {
+			log.Error("Failed to process resources validation response")
+		}
+	}
```
Comment (yahavi) [inline pipelines/services/validate.go:68]: You already check the status code in CheckResponseStatusWithBody. Please also consider nil before "CheckResponseStatusWithBody".

File: pipelines/services/validate.go [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+	return nil
+}
+
+// processValidatePipResourceResponse process validate pipeline resource response
+func processValidatePipResourceResponse(resp []byte) error {
+	validationResponse := make(map[string]ValidationResponse)
+	err := json.Unmarshal(resp, &validationResponse)
+	if err != nil {
+		return err
```
Comment (yahavi) [inline pipelines/services/validate.go:77]:
```suggestion
		return errorutils.CheckError(err)
```

File: pipelines/services/validate.go [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+	}
+	if len(validationResponse) == 0 {
+		return errors.New("pipelines not found")
```
Comment (yahavi) [inline pipelines/services/validate.go:80]:
```suggestion
		return errorutils.CheckErrorf("pipelines not found")
```

File: pipelines/services/validate.go [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+	}
+	for k, v := range validationResponse {
+		if v.IsValid != nil && *v.IsValid {
+			log.Info("Validation of pipeline resources completed successfully ")
```
Comment (yahavi) [inline pipelines/services/validate.go:84]:
```suggestion
			log.Info("Validation of pipeline resources completed successfully")
```

File: pipelines/services/validate.go [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+	}
+	return nil
+}
+
+// constructPipelinesURL creates URL with all required details to make api call
+// like headers, queryParams, apiPath
+func (vs *ValidateService) constructValidateAPIURL(qParams map[string]string, apiPath string) string {
+	uri, err := url.Parse(vs.ServiceDetails.GetUrl() + apiPath)
+	if err != nil {
+		log.Error("Failed to parse pipelines fetch run status url")
```
Comment (yahavi) [inline pipelines/services/validate.go:106]: Please return an error here

**Review by yahavi — CHANGES_REQUESTED** [review]

**Comment by eyalbe4** [comment]
Closing this PR for now.

**Pull request closed without merging**